### ISSUE TITLE
Proposal: Implement IEnumerable<int> or IEnumerable<long> for Range

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#462]
# Range API Proposal
## Implement `IEnumerable<int>` or `IEnumerable<long>` for Range

With above implementations, we can write:
```csharp
foreach (var i in [5..10]) Console.Write(i); // 5678910
[5..10].Count(); // 5
foreach (var i in [5..].Take(5)) Console.Write(i); // 5678910
[5..10].Where(i => (i & 1) == 1); // 5 7 9
......
```

**Comment by GrabYourPitchforks** [comment]
But the `[..^2]` in your stated example is a perfectly valid `Range`, and in fact it's an encouraged use case for easily trimming the last few pieces of data! Why would we want using such `Range` instances to result in NSE?

**Comment by hez2010** [comment]
> But the `[..^2]` in your stated example is a perfectly valid `Range`, and in fact it's an encouraged use case for easily trimming the last few pieces of data! Why would we want using such `Range` instances to result in NSE?

Consider `[..]` for an infinite number sequence, what will `[..^2]` like? 0~infinity-2?

**Comment by Gnbrkm41** [comment]
Given that `Index` can hold only up to `int.MaxValue`, why not `[0..int.MaxValue - 2]`?

**Comment by hez2010** [comment]
> Given that `Index` can hold only up to `int.MaxValue`, why not `[0..int.MaxValue - 2]`?
> 
> > IEnumerable for Range
> 
> For the same reason, I think it should be just `IEnumerable<int>`.

Sounds great. I've modified my proposal.

**Comment by GrabYourPitchforks** [comment]
My question was primarily to highlight that I don't think adding enumeration capabilities to the `Range` type makes sense, because we'd then define a special set of `Range` instances which are otherwise perfectly valid but where calling `GetEnumerator` has funky behavior. I see this work item as akin to https://github.com/dotnet/runtime/issues/417, where similar functionality is being proposed.

**Comment by GrabYourPitchforks** [comment]
Put another way, should the code samples below return the same value?

```cs
public int ComputeSumFoo(Range range, int[] array)
{
    int sum = 0;
    foreach (int idx in range) { sum += array[idx]; }
    return sum;
}

public int ComputeSumBar(Range range, int[] array)
{
    int sum = 0;
    foreach (int element in array[range]) { sum += element; }
    return sum;
}
```

**Comment by Gnbrkm41** [comment]
I agree. `Range` type really isn't meant to represent an arbitrary range of numbers, but rather a range of indices within a collection. While it can allow some cool code (`foreach (int i in 1..10)`) the "from end" bit can be hard to get right.

**Comment by bartonjs** [comment]
I don't think it's sound in the "from end" case, because that's a marker of "I need to know the total length".  If it doesn't throw in those cases I think it's an easy trap, and if it does throw then it's not a good candidate for IEnumerable... so I feel that this isn't a practical thing to do.

**Comment by reflectronic** [comment]
`Range` is [end-exclusive](https://docs.microsoft.com/en-us/dotnet/api/system.range.end?view=netcore-3.0#System_Range_End). The enumerator shouldn't yield the end value of the range (like it currently does in the proposal). That might make it more awkward to use, but this type is probably not the right place for this functionality anyway.

**Comment by GrabYourPitchforks** [comment]
If we wanted to support the syntax `foreach (int i in 1..10)` that would probably be a discussion for C# rather than the runtime team, especially since as mentioned earlier it would have different exclusive / inclusive considerations than today's `Range` type.

**Comment by jnm2** [comment]
@GrabYourPitchforks 

> But the `[..^2]` in your stated example is a perfectly valid `Range`, and in fact it's an encouraged use case for easily trimming the last few pieces of data! Why would we want using such `Range` instances to result in NSE?

At one stage this was the plan:

> # Enumerability
>
> `Range` should implement `IEnumerable<T>`, and needs to support the enumerable pattern of having a struct enumerator type. This is extra bulk in the type, but that's just how we do things.
>
> Enumeration on open ranges should probably throw.

https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-01-18.md#enumerability:

**Comment by scalablecory** [comment]
What's wrong with `Enumerable.Range` for this?

**Comment by wekempf** [comment]
Nothing wrong with Enumerable.Range, other than it doesn't have the nice range syntax. Personally, I like something like this.

```csharp
        static IEnumerable<int> AsEnumerable(this Range range, int length = int.MaxValue)
        {
            var (offset, len) = range.GetOffsetAndLength(length);
            return Enumerable.Range(offset, len);
        }
```

This gives you the following behaviors (note, there are some mistakes in the proposal text above, for example [5..10] can't both include 10 and return a Count() of 5).

```csharp
foreach (var i in (5..10).AsEnumerable()) Console.Write($"{i} "); // 5 6 7 8 9
(5..10).AsEnumerable().Count(); // 5
foreach (var i in (5..).AsEnumerable().Take(5)) Console.Write($"{i} "); // 5 6 7 8 9
(5..10).AsEnumerable().Where(i => (i & 1) == 1); // 5 7 9
foreach (var i in (..4).AsEnumerable()) Console.Write($"{i} ") // 0 1 2 3
foreach (var i in (..^2).AsEnumerable()) Console.Write($"{i} ") // 0 1 2 .. int.MaxValue-2
foreach (var i in (..).AsEnumerable()) Console.Write($"{i} ") // 0 1 2 .. int.MaxValue
foreach (var i in (..^2).AsEnumerable(10)) Console.Write($"{i} ") // 0 1 2 3 4 5 6 7
```

**Comment by khyperia** [comment]
> What's wrong with `Enumerable.Range` for this?

`Enumerable.Range` allocates, implementing IEnumerable for ranges (via struct iterator) would not. This is incredibly important for performance, and blocks a _lot_ of use cases.

Still truly baffled why this wasn't done, it was a very core original use case when proposing ranges in the first place - like, the number of people who responded to https://github.com/dotnet/csharplang/issues/2810 saying "just use ranges" is quite telling.
